File: backend/src/NLC.Core/Entities/DispatchNote.cs
using NLC.Core.Enums;

namespace NLC.Core.Entities;

public class DispatchNote
{
    public Guid           Id             { get; set; } = Guid.NewGuid();
    public Guid           JobId          { get; set; }
    public string         DnNumber       { get; set; } = default!;
    public string         CustomerName   { get; set; } = default!;
    public DispatchStatus DispatchStatus { get; set; } = DispatchStatus.PENDING;
    public DateTime       CreatedAt      { get; set; } = DateTime.UtcNow;
    public DateTime?      DispatchedAt   { get; set; }

    public JobCard                      JobCard  { get; set; } = default!;
    public ICollection<DispatchSkuLine> SkuLines { get; set; } = [];
}

public class DispatchSkuLine
{
    public Guid    Id               { get; set; } = Guid.NewGuid();
    public Guid    DnId             { get; set; }
    public string  SkuCode          { get; set; } = default!;
    public string  SkuDescription   { get; set; } = default!;
    public int     OrderedQty       { get; set; }
    public int     PickedQty        { get; set; }
    public int     DispatchedQty    { get; set; }
    public int     VarianceQty      => OrderedQty - DispatchedQty;
    public bool    VarianceApproved { get; set; }
    public string? ApprovedBy       { get; set; }

    public DispatchNote DispatchNote { get; set; } = default!;
}

File: backend/src/main/java/com/nlc/domain/entity/DispatchNote.java
package com.nlc.domain.entity;

import com.nlc.domain.enums.Enums.DispatchStatus;
import jakarta.persistence.*;
import lombok.*;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "dispatch_notes")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class DispatchNote {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "job_id", nullable = false)
    private UUID jobId;

    @Column(name = "dn_number", nullable = false, length = 50)
    private String dnNumber;

    @Column(name = "customer_name", nullable = false, length = 100)
    private String customerName;

    @Enumerated(EnumType.STRING)
    @Column(name = "dispatch_status", nullable = false, length = 20)
    @Builder.Default
    private DispatchStatus dispatchStatus = DispatchStatus.PENDING;

    @Column(name = "created_at", nullable = false)
    @Builder.Default
    private OffsetDateTime createdAt = OffsetDateTime.now();

    @Column(name = "dispatched_at")
    private OffsetDateTime dispatchedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "job_id", insertable = false, updatable = false)
    private JobCard jobCard;

    @OneToMany(mappedBy = "dispatchNote", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @Builder.Default
    private List<DispatchSkuLine> skuLines = new ArrayList<>();
}
